improvement(sdcm/tester.py): Add TEST_ERROR Argus status support

k0machi opened the pull request:
This commit adds additional logic into argus_finalize_test_run that
checks whether a failed run failed due to test error - e.g.
SpotTerminationEvent, where the framework encountered an environment
issue (not enough capacity) and cannot continue.

Task: scylladb/qa-tasks#1596

### Testing
<!-- Add links to Argus/Jenkins of test test done with this PR -->
<!-- This would help the reviewer to cross check what was tested, and and review the results as needed -->
- [x] [Argus](https://argus.scylladb.com/test/b0b22734-758e-47b4-b6ec-29e64d212d52/runs?additionalRuns[]=b923c361-6ea6-4d6c-83b0-ff652723e8a3)
- [x] [Jenkins](https://jenkins.scylladb.com/job/scylla-staging/job/alexey/job/alexey-argus-testing/95/)

### PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

fruch commented:
@k0machi 

here are some more example of events we should identify 
```
2024-01-18 03:24:21.252: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=42056688-4e0e-46f3-9246-6abf8ae8ddda, source=UpgradeTest.SetUp()
exception=403 FORBIDDEN QUOTA_EXCEEDED: Quota 'N2_CPUS' exceeded.  Limit: 3000.0 in region us-east1.
```

```
2024-06-06 20:59:42.137: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=a28d74ac-b721-41af-b9f0-10855b1eeeef, source=UpgradeTest.SetUp()
exception=An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient i4i.2xlarge capacity in the Availability Zone you requested (eu-west-1a). Our system will be working on provisioning additional capacity. You can currently get i4i.2xlarge capacity by not specifying an Availability Zone in your request or choosing eu-west-1b, eu-west-1c.
```

```
2024-06-06 21:06:51.282: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=82d8bbb6-14e5-49bf-8532-6679e9d9b6af, source=UpgradeTest.SetUp()
exception=An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient i4i.2xlarge capacity in zones with support for 'gp3' volumes. Our system will be working on provisioning additional capacity.
```

k0machi commented:
> @k0machi
> 
> here are some more example of events we should identify
> 
> ```
> 2024-01-18 03:24:21.252: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=42056688-4e0e-46f3-9246-6abf8ae8ddda, source=UpgradeTest.SetUp()
> exception=403 FORBIDDEN QUOTA_EXCEEDED: Quota 'N2_CPUS' exceeded.  Limit: 3000.0 in region us-east1.
> ```
> 
> ```
> 2024-06-06 20:59:42.137: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=a28d74ac-b721-41af-b9f0-10855b1eeeef, source=UpgradeTest.SetUp()
> exception=An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient i4i.2xlarge capacity in the Availability Zone you requested (eu-west-1a). Our system will be working on provisioning additional capacity. You can currently get i4i.2xlarge capacity by not specifying an Availability Zone in your request or choosing eu-west-1b, eu-west-1c.
> ```
> 
> ```
> 2024-06-06 21:06:51.282: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=82d8bbb6-14e5-49bf-8532-6679e9d9b6af, source=UpgradeTest.SetUp()
> exception=An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 4): We currently do not have sufficient i4i.2xlarge capacity in zones with support for 'gp3' volumes. Our system will be working on provisioning additional capacity.
> ```

Added, replaced matching by regex.

fruch reviewed (approved):
LGTM

soyacz commented on sdcm/tester.py:
```diff
@@ -559,14 +559,28 @@ def argus_get_scylla_version(self):
             severity=Severity.ERROR,
         )
 
+    def _is_test_error(self, events: dict[str, list[str]]) -> bool:
```
`_is_test_error` is very broad and vague, when reading SCT code, one must peek inside what it means.
Let's call it `_cloud_resources_error`

fruch replied:
I think the confusion here is that @k0machi was trying to name it after the state change it's gonna do.
i.e. the state is called 'Test-Error"

the state is vague and board, hence the function might be.

we could name it differently, but once we'll introduce something else into the fold.
like some failures of KMS at boot, it won't be resources anymore, we'll need to refactor/rename it.

maybe `should_we_change_test_state_to_test_error` :)

soyacz replied:
ok, but then let's have docstring explaining what we search for and update it when adding more checks

k0machi replied:
Added docstring

fruch commented:
@k0machi 

please add a docstring as was requested, and let's get this merged, and backported

k0machi commented:
> @k0machi
> 
> please add a docstring as was requested, and let's get this merged, and backported

Added docstring and also corrected logic a bit after testing it further.

soyacz reviewed (approved):
LGTM
